vsphere/resource_vsphere_license.go

```diff
@@ -0,0 +1,275 @@
+	ErrNoSuchKeyFound = errors.New("The key was not found")
+	// ErrKeyCannotBeDeleted is an error which occurs when a key that is used by VMs is
+	// being removed
+	ErrKeyCannotBeDeleted = errors.New("The key wasn't deleted")
+)
+
+func resourceVSphereLicense() *schema.Resource {
+
+	return &schema.Resource{
+
+		SchemaVersion: 1,
+
+		Create: resourceVSphereLicenseCreate,
+		Read:   resourceVSphereLicenseRead,
+		Update: resourceVSphereLicenseUpdate,
+		Delete: resourceVSphereLicenseDelete,
+
+		Schema: map[string]*schema.Schema{
+			"license_key": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"label": &schema.Schema{
+				Type:     schema.TypeList,
+				Optional: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"key": &schema.Schema{
+							Type:     schema.TypeString,
+							Required: true,
+							ForceNew: true,
+						},
+						"value": &schema.Schema{
+							Type:     schema.TypeString,
+							Required: true,
+						},
+					},
+				},
+			},
+
+			// computed properties returned by the API
+			"edition_key": &schema.Schema{
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"name": &schema.Schema{
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"total": &schema.Schema{
+				Type:     schema.TypeInt,
+				Computed: true,
+			},
+			"used": &schema.Schema{
+				Type:     schema.TypeInt,
+				Computed: true,
+			},
+		},
+	}
+}
+
+func resourceVSphereLicenseCreate(d *schema.ResourceData, meta interface{}) error {
+	log.Println("[INFO] Running the create method")
+
+	client := meta.(*govmomi.Client)
+	manager := license.NewManager(client.Client)
+
+	key := d.Get("license_key").(string)
+
+	log.Println(" [INFO] Reading the key from the resource data")
+	var finalLabels map[string]string
+	var err error
+	if labels, ok := d.GetOk("label"); ok {
+		finalLabels, err = labelsToMap(labels)
+		if err != nil {
+			// If errors are printed by terraform then no need to log errors here
+			return err
+		}
+	}
+	info, err := manager.Add(context.TODO(), key, finalLabels)
+	if err != nil {
+		return err
+	}
+
+	if err = DecodeError(info); err != nil {
+		return err
+	}
+
+	// This can be used in the read method to set the computed parameters
+	d.SetId(info.LicenseKey)
+
+	return resourceVSphereLicenseRead(d, meta)
+}
+
+func resourceVSphereLicenseRead(d *schema.ResourceData, meta interface{}) error {
+
+	log.Println("[INFO] Running the read method")
+
+	client := meta.(*govmomi.Client)
+	manager := license.NewManager(client.Client)
+
+	// the key was never present. Should set the ID to false.
+	if _, ok := d.GetOk("license_key"); !ok {
+		d.SetId("")
+		return nil
+	}
+
+	if info := getLicenseInfoFromKey(d.Get("license_key").(string), manager); info != nil {
+		log.Println("[INFO] Setting the values")
+		d.Set("edition_key", info.EditionKey)
+		d.Set("total", info.Total)
+		d.Set("used", info.Used)
+		d.Set("name", info.Name)
+
```

Comment:
If i update the labels here, then the state provided by terraform files and terraform state stored will drift apart, for example if server has labels `x:y` and local state shows it as `x:yy` then the server state has to be changed not the other way around. That's why only computed properties are set as they are the one which the user is not responsible for.